File: src/main/java/net/dblsaiko/illuminate/client/LightContainer.java
package net.dblsaiko.illuminate.client;

import com.mojang.blaze3d.platform.GlStateManager;
import net.dblsaiko.illuminate.client.api.Light;
import org.joml.Matrix4f;
import org.joml.Matrix4fc;
import org.joml.Vector3f;
import org.joml.Vector3fc;
import org.lwjgl.opengl.GL31;

public class LightContainer {
    private final Light light;
    private int depthTex = 0;

    private final Vector3f cachedPos = new Vector3f();
    private float cachedYaw;
    private float cachedPitch;
    private float cachedRoll;
    private float cachedFov;
    private float cachedAspect;
    private float cachedNear;
    private float cachedFar;

    private final Matrix4f mvp = new Matrix4f();
    private boolean mvpDirty = true;
    private final Matrix4f mv = new Matrix4f();
    private boolean mvDirty = true;
    private final Matrix4f p = new Matrix4f();
    private boolean pDirty = true;

    public LightContainer(Light light) {
        this.light = light;
    }

    public Light light() {
        return this.light;
    }

    public int depthTex() {
        if (this.depthTex == 0) {
            this.depthTex = GlStateManager._genTexture();
            GlStateManager._bindTexture(this.depthTex);
            GlStateManager._texParameter(GL31.GL_TEXTURE_2D, GL31.GL_TEXTURE_MAG_FILTER, GL31.GL_LINEAR);
            GlStateManager._texParameter(GL31.GL_TEXTURE_2D, GL31.GL_TEXTURE_MIN_FILTER, GL31.GL_LINEAR);
            GlStateManager._texParameter(GL31.GL_TEXTURE_2D, GL31.GL_TEXTURE_WRAP_S, GL31.GL_CLAMP_TO_EDGE);
            GlStateManager._texParameter(GL31.GL_TEXTURE_2D, GL31.GL_TEXTURE_WRAP_T, GL31.GL_CLAMP_TO_EDGE);
        }

        return this.depthTex;
    }

    public Matrix4fc mvp() {
        this.update();

        Matrix4fc mv = this._mv();
        Matrix4fc p = this._p();

        if (this.mvpDirty) {
            p.mul(mv.translate(this.cachedPos.negate(new Vector3f()), this.mvp), this.mvp);
            this.mvpDirty = false;
        }

        return this.mvp;
    }

    public Matrix4fc mv() {
        this.update();
        return this._mv();
    }

    public Matrix4fc p() {
        this.update();
        return this._p();
    }

    private Matrix4fc _mv() {
        if (this.mvDirty) {
            this.mv.set(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1);
            this.mv.rotate(this.cachedRoll, 0, 0, 1);
            this.mv.rotate(this.cachedPitch, 1, 0, 0);
            this.mv.rotate(this.cachedYaw, 0, 1, 0);
            this.mvDirty = false;
            this.mvpDirty = true;
        }

        return this.mv;
    }

    private Matrix4fc _p() {
        if (this.pDirty) {
            this.p.set(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1);
            this.p.perspective(this.cachedFov, this.cachedAspect, this.cachedNear, this.cachedFar);
            this.pDirty = false;
            this.mvpDirty = true;
        }

        return this.p;
    }

    public void update() {
        Vector3fc pos = this.light.pos();
        float yaw = this.light.yaw();
        float pitch = this.light.pitch();
        float roll = this.light.roll();
        float fov = this.light.fov();
        float aspect = this.light.aspect();
        float near = this.light.near();
        float far = this.light.far();

        if (!this.cachedPos.equals(pos)) {
            this.cachedPos.set(pos);
            this.mvpDirty = true;
        }

        if (this.cachedYaw != yaw) {
            this.cachedYaw = yaw;
            this.mvDirty = true;
        }

        if (this.cachedPitch != pitch) {
            this.cachedPitch = pitch;
            this.mvDirty = true;
        }

        if (this.cachedRoll != roll) {
            this.cachedRoll = roll;
            this.mvDirty = true;
        }

        if (this.cachedFov != fov) {
            this.cachedFov = fov;
            this.pDirty = true;
        }

        if (this.cachedAspect != aspect) {
            this.cachedAspect = aspect;
            this.pDirty = true;
        }

        if (this.cachedNear != near) {
            this.cachedNear = near;
            this.pDirty = true;
        }

        if (this.cachedFar != far) {
            this.cachedFar = far;
            this.pDirty = true;
        }
    }

    public void destroy() {
        if (this.depthTex != 0) {
            GlStateManager._deleteTexture(this.depthTex);
        }

        this.depthTex = 0;
    }
}

File: src/main/kotlin/client/render/LightContainer.kt
package therealfarfetchd.illuminate.client.render

import org.joml.Matrix4f
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL12
import therealfarfetchd.illuminate.client.api.Light
import therealfarfetchd.illuminate.client.glwrap.WGlTexture2D
import therealfarfetchd.illuminate.common.util.ext.perspective
import therealfarfetchd.illuminate.common.util.ext.times
import therealfarfetchd.illuminate.common.util.ext.unaryMinus

data class LightContainer(val light: Light) {

  val depth = WGlTexture2D.create()

  init {
    depth.texParameter(GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_LINEAR)
    depth.texParameter(GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_LINEAR)
    depth.texParameter(GL11.GL_TEXTURE_WRAP_S, GL12.GL_CLAMP_TO_EDGE)
    depth.texParameter(GL11.GL_TEXTURE_WRAP_T, GL12.GL_CLAMP_TO_EDGE)
  }

  private var cachedPos = light.pos
    set(value) {
      mvpDirty = mvpDirty || field != value
      field = value
    }

  private var cachedYaw = light.yaw
    set(value) {
      mvDirty = mvDirty || field != value
      field = value
    }

  private var cachedPitch = light.pitch
    set(value) {
      mvDirty = mvDirty || field != value
      field = value
    }

  private var cachedRoll = light.roll
    set(value) {
      mvDirty = mvDirty || field != value
      field = value
    }

  private var cachedFov = light.fov
    set(value) {
      pDirty = pDirty || field != value
      field = value
    }

  private var cachedAspect = light.aspect
    set(value) {
      pDirty = pDirty || field != value
      field = value
    }

  private var cachedNear = light.near
    set(value) {
      pDirty = pDirty || field != value
      field = value
    }

  private var cachedFar = light.far
    set(value) {
      pDirty = pDirty || field != value
      field = value
    }

  fun update() {
    cachedPos = light.pos
    cachedYaw = light.yaw
    cachedPitch = light.pitch
    cachedRoll = light.roll
    cachedFov = light.fov
    cachedAspect = light.aspect
    cachedNear = light.near
    cachedFar = light.far
  }

  private var mvpDirty = true

  private var mvDirty = true
    set(value) {
      field = value
      if (value) mvpDirty = true
    }

  private var pDirty = true
    set(value) {
      field = value
      if (value) mvpDirty = true
    }

  var mv: Matrix4f = Matrix4f()
    get() {
      update()
      if (mvDirty) {
        field = Matrix4f()
          .rotate(0f, 0f, 1f, cachedRoll)
          .rotate(1f, 0f, 0f, cachedPitch)
          .rotate(0f, 1f, 0f, cachedYaw)
        mvDirty = false
      }
      return field
    }
    private set

  var p: Matrix4f = Matrix4f()
    get() {
      update()
      if (pDirty) {
        field = perspective(cachedFov, cachedAspect, cachedNear, cachedFar)
        pDirty = false
      }
      return field
    }
    private set

  var mvp: Matrix4f = Matrix4f()
    get() {
      update()
      if (mvpDirty) {
        field = p.mul(mv.translate(-cachedPos))
        mvpDirty = false
      }
      return field
    }
    private set

  fun destroy() {
    depth.destroy()
  }

}
